massifs/localmassifreader.go

```diff
@@ -0,0 +1,384 @@
+func (r *LocalReader) GetMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint64,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+
+	// short circuit direct match, regardless of mode, to support explicit paths
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, massifIndex))
+	}
+
+	directory, err := r.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+
+	return copyCachedMassifOrErr(r.cache.ReadMassif(directory, massifIndex))
+}
+
+// GetSeal reads the seal identified by the tenant identity and massif index
+func (r *LocalReader) GetSeal(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint64,
+	opts ...ReaderOption,
+) (SealedState, error) {
+
+	// short circuit direct match, regardless of mode, to support explicit paths
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedSealOrErr(dirEntry.ReadSeal(r.cache, tenantIdentityOrLocalPath))
+	}
+
+	directory, err := r.ResolveSealDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return SealedState{}, err
+	}
+
+	return copyCachedSealOrErr(r.cache.ReadSeal(directory, massifIndex))
+}
+
+// GetSignedRoot satisfies the SealGetter interface.
+// This is the default seal getter for the local reader when GetVerifiedContext is called.
+func (r *LocalReader) GetSignedRoot(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint32,
+	opts ...ReaderOption,
+) (*cose.CoseSign1Message, MMRState, error) {
+	sealedState, err := r.GetSeal(ctx, tenantIdentityOrLocalPath, uint64(massifIndex), opts...)
+	if err != nil {
+		return nil, MMRState{}, err
+	}
+	return &sealedState.Sign1Message, sealedState.MMRState, nil
+}
+
+// GetHeadMassif reads the most recent massif in the log identified by the tenant identity
+func (r *LocalReader) GetHeadMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+	}
+
+	directory, err := r.cache.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	err = r.cache.FindMassifFiles(directory)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	dirEntry, ok = r.cache.GetEntry(directory)
+	if !ok {
+		// It's a bug in FindMassifFiles if this case actually happens in practice.
+		return MassifContext{}, fmt.Errorf("failed to prime directory: %s", directory)
+	}
+	return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+}
+
+func (r *LocalReader) GetFirstMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
```

Comment:
why not call `Getmassif(ctx, tenantIdentityOrLocalPath, 0, opts)`?

Reply:
I will try and re-factor that more if I can. The chief difficulty was that GetFirstMassif triggers a search, but GetMassif must fail if the requested one doesn't exist.

Reply:
Oh wait, yeah. there is more to this. FirstMassifIndex is not necessarily massif 0. It's the massif log file with the lowest index that is found locally. And that is only known by first scaning the directory. Some of the later improvements have made it possible to simplify both GetMassif and this method so I will endevour to do that and hope it makes things clearer

Reply:
Right, that was worth another look. I've managed to remove quite a lot of the complexity from these methods. The logdircache got improved and simplified, but i hadn't carried that all the way through.